spring-data-neo4j-rx/src/main/java/org/neo4j/springframework/data/core/cypher/Case.java

```diff
@@ -0,0 +1,233 @@
+ * @since 1.0
+ */
+@API(status = EXPERIMENTAL, since = "1.0")
+public abstract class Case implements Visitable {
+
+	private CaseElse caseElse;
+	private final List<CaseWhenThen> caseWhenThens;
+
+	static Case create(Expression caseExpression) {
+		return new SimpleCase(caseExpression, new ArrayList<>());
+	}
+
+	static Case create() {
+		return new GenericCase(new ArrayList<>());
+	}
+
+	protected Case(List<CaseWhenThen> caseWhenThens) {
+		this.caseWhenThens = caseWhenThens;
+	}
+
+	@Nullable
+	abstract Expression getCaseExpression();
+
+	void setCaseElse(CaseElse caseElse) {
+		this.caseElse = caseElse;
+	}
+
+	List<CaseWhenThen> getCaseWhenThens() {
+		return caseWhenThens;
+	}
+
+	public OngoingWhenThen when(Expression whenExpression) {
+		return new OngoingWhenThen(whenExpression);
+	}
+
+	/**
+	 * Specification for a renderable, complete CASE statement
+	 */
+	public interface CaseEnding extends Condition {
+		OngoingWhenThen when(Expression whenExpression);
+
+		CaseEnding elseDefault(Expression elseExpression);
+
+	}
+
+	/**
+	 * Special implementation of the {@link Case} class to support simple case with an initial expression / condition.
+	 */
+	@API(status = EXPERIMENTAL, since = "1.0")
+	public static class SimpleCase extends Case {
+
+		private final Expression caseExpression;
+
+		public SimpleCase(Expression caseExpression, List<CaseWhenThen> caseWhenThens) {
+			super(caseWhenThens);
+			this.caseExpression = caseExpression;
+		}
+
+		@Override
+		Expression getCaseExpression() {
+			return caseExpression;
+		}
+
+		/**
+		 * The renderable implementation of {@link SimpleCase}.
+		 */
+		public final static class EndingSimpleCase extends SimpleCase implements CaseEnding {
+
+			public EndingSimpleCase(Expression caseExpression, List<CaseWhenThen> caseWhenThens) {
+				super(caseExpression, caseWhenThens);
+			}
+
+			public CaseEnding elseDefault(Expression elseExpression) {
+				this.setCaseElse(new CaseElse(elseExpression));
+				return this;
+			}
+		}
+	}
+
+	/**
+	 * Implementation of the {@link Case} class to support generic case.
+	 */
+	@API(status = EXPERIMENTAL, since = "1.0")
+	public static class GenericCase extends Case {
+
+		public GenericCase(List<CaseWhenThen> caseWhenThens) {
+			super(caseWhenThens);
+		}
+
+		@Override
+		Expression getCaseExpression() {
+			return null;
+		}
+
+		/**
+		 * The renderable implementation of {@link GenericCase}.
+		 */
+		public final static class EndingGenericCase extends GenericCase implements CaseEnding {
+
+			public EndingGenericCase(List<CaseWhenThen> caseWhenThens) {
+				super(caseWhenThens);
+			}
+			public CaseEnding elseDefault(Expression elseExpression) {
+				this.setCaseElse(new CaseElse(elseExpression));
+				return this;
+			}
+		}
+	}
+
+	@Override
+	public void accept(Visitor visitor) {
+		visitor.enter(this);
+		if (getCaseExpression() != null) {
+			getCaseExpression().accept(visitor);
+		}
+		for (CaseWhenThen caseWhenThen : getCaseWhenThens()) {
```

Comment:
I used the `.forEach(c -> c.accept(visitor));` form in other similar places. No preference here, but unity :)